### ISSUE TITLE
Keyboard stutters

### ISSUE THREAD

**Issue by ghost** [input-leap/input-leap#269]
> This issue has been migrated from old Barrier Github repository https://github.com/debauchee/barrier/issues/269
> 
> Issue created on: 2019-03-14 by @AverageGuy
> Issue last updated on: 2020-07-30

### Operating Systems ###

Server: Mint Mate 18
Client: Mint Mate 19

### Barrier Version ###

1.9.0

### Steps to reproduce bug ###
Start typing, sooner or latter you will type a letter, number, punctuation or space and magically 2 or more of them will appear.  It's fairly intermittent.  I typed only one t in "later", for instance.  

### Other info ###

* When did the problem start to occur? After I installeddd barrier on both systems from source.
* Is there a way to work around it? No, other than using the backspace key.
* Does this bug prevent you from using Barrier entirely? No
Put anything else you can think of here.
It affects anything you aree typing into.  This bug, a terminal window, Vim instance, are exampllles of where it happens, and no I'm not holding the key down.  I'm a touch typiist and get on/off the keys pretty fast.

---

> Commented on: 2019-03-14 by @AverageGuy

All works OK on the server.  No stuttering.  I replaced synergy with barrieer but never saw it with synerrgy.  The client is a new machine running Mate 19 annd I was unable to get synergy to run on it.  Incompatible versions it said.  That's why I am runniiing barrier....
I didn't type but onnnnne period.

---

> Commented on: 2019-03-15 by @AdrianKoshka

Are you trying to use barrier over wifi?

---

> Commented on: 2019-03-17 by @AverageGuy

Both systems are connected to a gigabit swiiiitch via ethernet.

---

> Commented on: 2019-10-09 by @AverageGuy

At a different location, I have the same issue.  Mint 19 on one machine, Mint 18 on the other.  The only difference is the master is running 19 at this location.  The original problem exists when the master is on 18.  Stuttering is on the slave in both cases.

---

> Commented on: 2019-12-02 by @charliehorse55

I also have this problem. I have a VFIO setup, and I use barrier to share the mouse between OSes. The server runs on linux, and the guest is either windows 10 or another linux instance. The stuttering is ONLY present on the linux guest, windows works fine. In both cases, the machines are connected through a virtualized 10gbit link, so I highly doubt this is a networking issue. 

---

> Commented on: 2020-01-27 by @charliehorse55

Does anyone have any information on this? The stutters can be quite long for me (several seconds). If I happened to be pressing a key while editing code, it will mess up what I'm editing. This happens multiple times per hour. Luckily I can ctrl-z, but it breaks my flow. 

It's really funny when it gets stuck on backspace and deletes 10+ lines of code :)

Is there a bug bounty for this repo anywhere? I could easily donate $100 for this to be fixed!

---

> Commented on: 2020-05-29 by @amerlyq

TL;DR: use `xset r off` on client.

BET:DEV: fix barrier(1) client code to disable autorepeat on XTest simulated keys (only simulated by it itself).

I use latest version barrier with `ArchLinux PC -> ArchLinux Laptop` over home router 5G wifi.
I strongly suspect the root problem lays in "X11 autorepeat" feature working *both* on server and client.
Barrier sends keydown and keyup events separately, and if network latency becomes more than autorepeat trigger on client -- it will result in wrong autorepeat. It's easy to verify: 
* use `xset r rate 80 20` on client to observe the issue on almost each keypress
* use `xset r off` -- and issue on client is gone.

There is no need to disable autorepeat on server -- its events will propagate to client and "simulate" autorepeat there. If they are not propagated (another bug?) then you can increase autorepeat rate to be above your network latency: `xset r rate 800 50`

---

> Commented on: 2020-05-29 by @charliehorse55

Unfortunately, that's not the problem. The mouse always freezes along with the keyboard input. 

---

> Commented on: 2020-05-29 by @p12tic

This bug looks like a duplicate of #617 which is fixed in #679.

---

> Commented on: 2020-07-30 by @deters

Probably wifi lag... barrier could reduce this by implementing a "keep_alive" option: sending dummy packets at a constant rate to the client in order to keep wifi module active. 

My current workaround:

starting netcat in the server: 
nc -l 5555

using netcat in the client to send a "." to the server every 50ms:
while [ 1 ]; do sleep 0.05; echo .; done | nc 192.168.1.190 5555
